src/Microsoft.AspNet.Hosting/HostingEngine.cs

```diff
@@ -105,9 +106,16 @@ private void EnsureApplicationStartup(HostingContext context)
                 return;
             }
 
-            context.ApplicationStartup = _startupManager.LoadStartup(
+            var diagnosticMessages = new List<string>();
+            context.ApplicationStartup = _startupLoader.LoadStartup(
                 context.ApplicationName,
-                context.EnvironmentName);
+                context.EnvironmentName,
+                diagnosticMessages);
+
+            if (context.ApplicationStartup == null)
+            {
+                throw new Exception(diagnosticMessages.Aggregate("TODO: web application entrypoint not found message", (a, b) => a + "\r\n" + b));
```

Comment:
Use appropriate exception type. This is probably an ArgumentException, with `context` being the invalid argument.

Also, no more "TODO" in the error message - just put whatever the error message should be.